# Submodule 1: Introduction to Genome Sequencing and Assembly
--------
## Overview
Genomics is the comprehensive study of an organism's complete set of DNA, including all of its genes. It provides a deep understanding of the genetic blueprint that governs the biology, function, and behavior of organisms. With the advent of advanced sequencing technologies, genomics has become an essential tool for exploring the genetic basis of health and disease, understanding evolution, and studying biodiversity. In this submodule you will learn how DNA and RNA sequencing data is produced and how to reconstruct the genome of an organism from short DNA sequences generated by sequencing technologies. The process will be completed using through an interactive markdown document and will use command-line bioinformatic tools. We will explore the key concepts and methodologies in genomics, including how to assemble a genome, how to assess its quality, how to annotate the genome, and how to perform comparative analyses.

<p align="center">
  <img src="images/diagram-WGS.png" width="50%"/>
</p>

Figure 1. Overview of the bioinformatic workflow for the four submodules in this tutorial.


### Learning Objectives:

- **Understand how sequencing data is generated**: Learn the techniques and technologies used to generate genomic sequencing data, including next-generation sequencing (NGS) and its impact on genome assembly.

- **Develop an understanding of core bioinformatic input/output formats in genomics**: Gain proficiency in handling common file formats like FASTQ, BAM, and GFF, and understand their role in processing, storing, and analyzing genomic sequencing data.

- **Acquire the skills to assemble raw sequencing reads into a draft genome**: Master the process of genome assembly, from quality control of sequencing data to using assembly tools for constructing a draft genome from raw reads.

## Background: How is sequencing data produced?

### What is Next-Generation Sequencing (NGS)?

Next-Generation Sequencing (NGS) is a high-throughput method that allows for rapid sequencing of DNA. Illumina remains the most widely used sequencing platform. It is accurate, inexpensive, and fast, providing up to 540 Gb of data in about a 1-2 days. Other long-read sequencing technologies (i.e. PacBio and Nanopore) are becoming widely used in the research community, and will be discussed in later bioinformatic modules.

Figure 2 outlines the basic workflow for sequencing a microbial genome. It begins with the **isolation of bacterial cultures**, which involves culturing bacteria from a sample on selective media to obtain a pure strain. Once a pure culture is established, **DNA extraction** is performed to isolate the genomic DNA from the bacterial cells.

For Illumina-based sequencing, DNA is **fragmented** into smaller pieces, and adapters and barcode indexes are added to each fragment to enable sequencing and sample multiplexing. This step is called **library preparation**. The prepared sequencing libraries are then attached to a specially coated surface on a microscope slide, forming dense clusters of identical DNA fragments ready for sequencing. IN a process called **sequencing by synthesis**, fluorescently labeled nucleotides are incorporated into the clusters one at a time, using the complementary base pairing properties of DNA. Each incorporated nucleotide emits a distinct signal, allowing the sequencer to determine the DNA sequence in real time.


This technology enables millions of fragments to be sequenced simultaneously, significantly increasing throughput and reducing costs.


Isolate Bacteria            |  Extract DNA
:-------------------------:|:-------------------------:
![alt text](https://github.com/Joseph7e/HCGS-Genomics-Tutorial/blob/master/petri.jpg?raw=true)  |  <img src="https://www.cephamls.com/wp-content/uploads/2019/02/DNA-Extraction-Figure-3-22.jpg" width="420">

Prepare Library           |  Sequence DNA
:-------------------------:|:-------------------------:
<img src="https://jef.works//assets/blog/librarystructure.png" width="520">  |  <img src="https://github.com/Joseph7e/HCGS-Genomics-Tutorial/blob/master/hiseq.png?raw=true" width="320">

Figure 2. The process of sequencing an organisms DNA.


## Video Lesson: How next-generation sequencing works.

The following video describes the process of preparing an Illumina sequenicng library and how genome sequenicng with an Illumina sequencer works. We also compare and contrast this to other sequencing technologies (PacBio and Nanopore).

[![Sequencing by Synthesis](images/cluster-generation.jpg)](https://www.youtube.com/watch?v=p4vKJJlNTKA)

In [2]:
%%html
<iframe width="560" height="315" src="https://www.youtube.com/embed/p4vKJJlNTKA" frameborder="0" allowfullscreen></iframe>

## **Install required software**

Four main tools are required for for Submodule 1: **fastq-dump, fastqc, fastp, and spades**. Tools are preinstalled into a docker image for this tutorial, but we will demonstrate how to install these tools using __[Conda](https://docs.conda.io/en/latest/)__, an open-source package management system and environment management tool. Conda helps to easily install, update, and manage software dependencies across different platforms. When we install the tools we select specific versions to ensure consistent behavior and reproducibility of results across different environments and systems.

### List of software

| **Tool**       | **Description**                                                                                                                                                           |
|:---------------|:--------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| **sra-tools**      | fastq-dump is used to retrieve publicly available data from the NCBI Sequence Read Archive (SRA).                                                 |
| **fastqc**        | Used to assess the quality of raw sequencing data by generating visual reports on metrics like read length, GC content, and sequence duplication.|
| **fastp**   | Used for quality control and pre-processing of FASTQ files, including read trimming, adapter removal, and filtering out low-quality reads.                          |
| **spades**      | Used for genome assembly, creating contigs from high-quality, pre-processed sequencing reads.                  |  

<div class="alert alert-block alert-warning"> <b>Attention:</b> The code below is written in Bash. The <code>%%bash</code> at the beginning of the cell instructs the Jupyter notebook to execute the code as a Bash script when the cell is run. This is the same code you would run from a standard command-line terminal</a>. </div>

<div class="alert alert-block alert-info"><b>Tip</b>: To execute the code you can either press **ctrl-ENTER** when the cell is selected or by pressing the play button at the top of the notebook</div>

In [26]:
%%bash

echo -e "\033[1mNote: All command-line tools are preinstalled in a docker image\033[0m"
echo -e "\033[1mNote: The command below provides an alternative means to aquire the needed tools using conda, but is not run by default\033[0m"
echo "------------------------------------------------------------------------------"
# Install tools using mamba (a conda alternative) with specified versions

echo "mamba install --channel bioconda \
    python=3.9 \
    sra-tools=3.1.1 \
    fastqc=0.11.8 \
    fastp=0.23.4 \
    spades=4.0.0 \
    -y > /dev/null 2>&1"

echo "------------------------------------------------------------------------------"
echo -e "\033[1mInstallation of sra-tools, fastqc, fastp, and spades complete.\nDisplaying versions:\033[0m"

# Confirm installation by checking versions
fastq-dump --version | tr -d '\n'
fastqc --version
fastp --version
spades.py --version

Note: All command-line tools are preinstalled in a docker image
Note: The command below provides an alternative means to aquire the needed tools using conda, but is not run by default
------------------------------------------------------------------------------
mamba install --channel bioconda     python=3.9     sra-tools=3.1.1     fastqc=0.11.8     fastp=0.23.4     spades=4.0.0     -y > /dev/null 2>&1
------------------------------------------------------------------------------
Installation of sra-tools, fastqc, fastp, and spades complete.
Displaying versions:
fastq-dump : 3.1.1FastQC v0.12.1


fastp 0.23.4


SPAdes genome assembler v4.0.0


<div class= "alert alert-block alert-info"><b>Tip</b>: use <code>\</code> to break a long command into multiple lines</div>

<div class= "alert alert-block alert-info"><b>Tip</b>: use <code>/dev/null 2>&1</code> to write standard error and standard output to <i>/dev/null</i>, effectively discarding them so your screen doesn't get flooded with text.</div></div>

## **Starting data**
The data used for this module is described in a manuscript comparing phenotypic and whole-genome sequencing derived AMR profiles (Painset et al. 2020) and is available under the SRA accession SRR10056829. Genomic DNA was extracted, fragmented and prepared for multiplexing using a **Nextera XT DNA Sample Preperation Kits**. Paired-end sequencing data was produced on an **Illumina HiSeq 2500** platform.

The manuscript includes sequencing read datasets for 528 isolates of *Campylobacter* spp. (452 *C. jejuni* and 76 *C. coli*) from human (494), food (21) and environmental (2) sources. We will start by walking through the process on just a single dataset. We will later expand this to include a larger set of this data.

In [17]:
%%bash

# Capture the SRA accession in a variable 
accession=SRR10056829

# Download data from the SRA using the variable from above 

# prefetching downloads metadata of the SRA records (5.629s)
echo "prefetching metadata from the NCBI SRA"
prefetch -v $accession > /dev/null 2>&1

# fastq-dump downlaods the reads and compresses them using standard gzip compression. (1m33.96s)
echo "downlaoding the reads and compressing them using standard gzip compression"
fastq-dump --outdir raw-reads --gzip --split-files "$accession"/"$accession".sra > /dev/null 2>&1

# Remove the prefetch directory
rm -r $accession

echo Process Complete

prefetching downloads metadata of the SRA records
downlaoding the reads and compressing them using standard gzip compression
Process Complete


<div class= "alert alert-block alert-info"><b>Tip</b>: Assigning certain values to variables (like the accession above) allows you to easily adjust input to a set of commands and reduces the amount of typing for long or complicated file names</div>

## FASTQ file format
The data we just downloaded are in **FASTQ** file format. Let's examine this data and file format.

Let's run three sections of code to learn about the FASTQ file format. Please also see a detailed description of FASTQ files in the NIGMS module [Fundementals of Bionformatics](https://github.com/NIGMS/Fundamentals-of-Bioinformatics)

In [27]:
%%bash

# If the above code worked you should have two FASTQ files in a directory called raw-reads
ls raw-reads/

SRR10056829_1.fastq.gz
SRR10056829_2.fastq.gz


<div class="alert alert-block alert-warning"> <b>Attention:</b> Notice the the files above have the .gz extension. FASTQ data is commonly gzip compressed to save sapce. Most bioinformatic tools accept or require gzipped FASTQs. </div>

In [28]:
%%bash

# define variable names for reads
forward=raw-reads/SRR10056829_1.fastq.gz
reverse=raw-reads/SRR10056829_2.fastq.gz

# view the first four lines
echo "-------- FASTQ READ ---------"
zcat $forward | head -n 4
echo "-----------------------------"

-------- FASTQ READ ---------
@SRR10056829.1 1 length=100
NCCCCAAGGAATAACATTCATAACCCCCATAGCCGAAGTAATAATCAATAACATTGAAGTCTTTCTTATACCTAAACGCTCATAAATAGGTAATAATGCA
+SRR10056829.1 1 length=100
#0<FFFFFFFFFFIIFFIIIIIIFIIIIIIIIIFIFFIIIIIIIIFFFIIIIIIIIFIIIIIIIIIIFFFFFFFBFFFFFFFFFFFFFBFBBFFFFFFFF
-----------------------------


### Explanation:
The above code should have displayed the first four lines of the FASTQ file. **Each sequencing read in a FASTQ formatted file is made up of four lines**. 

The length of the sequencing reads can vary depending on the type of sequencing instrument and flow cell used, but they should be consistent across the raw data. In our example datset the reads are 100 bps in length. Illumina sequencers produce reads with length lengths randing between 50 and 300 bps. Run the following code to print the length of the reads.

In [29]:
%%bash
zcat raw-reads/SRR10056829_1.fastq.gz | head -n 2 | tail -n 1 | wc -c

101


<div class= "alert alert-block alert-info"><b>Tip</b>: The number above is one higher than the actual read length, this is because of the newline charcater at the end of each line that counted with the wc command</div>

## Do we have enough data? 

Most bacterial genomes consist of a single, circular chromosome, although some bacteria may have linear chromosomes or multiple chromosomes. In addition to the main chromosome, plasmids (small circular DNA molecules) often coexist, carrying additional genes. Compared to eukaryotes, bacterial genomes are typically much smaller, ranging from approximately 0.5 to 10 megabases (Mb) in size. For instance, Escherichia coli has a genome size of about 4.6 Mb. This genome can be represented as a string of 4.6 million nucleotides—comprising "A"s, "T"s, "C"s, and "G"s.

| Organism                        | Genome Size (Mb) |
|----------------------------------|------------------|
| *Escherichia coli* (Bacteria)    | 4.6              |
| *Streptomyces coelicolor* (Bacteria) | 8.7          |
| *Mycoplasma genitalium* (Bacteria)   | 0.58         |
| *Homo sapiens* (Human)           | 3,200            |
| *Drosophila melanogaster* (Fruit Fly) | 139          |
| *Arabidopsis thaliana* (Thale Cress) | 135           |
| *Saccharomyces cerevisiae* (Yeast)   | 12.1         |
| *Caenorhabditis elegans* (Nematode)  | 100           |
| *Mus musculus* (Mouse)           | 2,700            |
| *Zea mays* (Maize)               | 2,300            |


### What is sequencing 'coverage'?
**Breadth of coverage** refers to the percentage of the genome that has been sequenced at least once. It tells you how much of the genome has any sequencing data aligned to it, but it doesn’t indicate how many times each base has been sequenced. High breadth of coverage means most or all regions of the genome are represented in the sequencing data.

**Depth of coverage**, on the other hand, refers to the average number of times each base in the genome is sequenced, often represented as "X" (e.g., 30x depth). Depth is important for accuracy, as higher depth ensures that individual bases are sequenced multiple times, improving confidence in variant calling and reducing sequencing errors.

When researchers say *coverage* they usually mean *depth of coverage*.


The equation for genome depth of coverage is:

$$
\text{Coverage} = \frac{N \times L}{G}
$$

Where:
- \( N \) = Number of reads
- \( L \) = Read length
- \( G \) = Genome size


The total number of bases sequenced, the numerator in the above equation, is also called the total **throughput** for the sample.


### How can we determine our total throughput and coverage?

Lets do the math based on the data we have so far, we know the read length, and we can use a best guess at the genome size as 5 Mb for now. So lets count our number of reads.

In [32]:
%%bash

# counting the lines and dividing by 4. Remember each read entry is exactly four lines long. These numbers should match.
echo "The total number of lines in the fastq file:"
zcat raw-reads/*_1.fastq.gz | wc -l

The total number of lines in the fastq file
6365120


In [36]:
%%bash

# we can do this calculation from the terminal with echo and bc (bc is the terminal calculator)
num_reads=6365120/4
read_length=100
# adjust this number to the organisms genome size.
genome_size=5000000

# calculate our total throughput
echo "Total number of base pairs sequenced:"
#throughput=$(echo $num_reads " * " $read_length " * 2" | bc)
throughput=$(awk "BEGIN {print $num_reads * $read_length * 2}")
echo $throughput

# calculate our estimated coverage
echo "The estimated depth of coverage for this genome dataset:"
#coverage=$($throughput " / " $genome_size | bc)
coverage=$(awk "BEGIN {print $throughput / $genome_size}")
echo $coverage


Total number of base pairs sequenced:
318256000
The estimated depth of coverage for this genome dataset:
63.6512


### Explanation

Our estimated coverage, assuming a 5 Mb genome, is 63X. This means we are estimated to have sequenced every base in the genome approximately 63 times.

**Coverage and De Novo Assembly:**
For de novo genome assembly, coverage is a critical factor. Typically, **30X to 50X coverage** is recommended for most bacterial genomes, as it provides enough data to accurately assemble the genome without over-assembling, which could lead to potential issues. However, the required coverage can vary depending on factors such as the organism’s genome size, complexity, and the sequencing platform used.

- **Bacterial genomes**: 30X to 50X is generally sufficient.
- **Eukaryotic genomes**: Larger genomes may require higher coverage, often in the range of 50X to 100X or more, depending on the complexity and repetitive regions.

Higher coverage helps ensure the assembly is accurate and complete, reducing the chances of gaps or errors. However, after a certain point, increasing coverage does not significantly improve the assembly quality and only increases sequencing costs.


## Step 1: Quality Assessment with FASTQC
- **Program**: **FASTQC**
- **Citation**: *Andrews, S. (2010). FastQC: A Quality Control Tool for High Throughput Sequence Data. Available online: http://www.bioinformatics.babraham.ac.uk/projects/fastqc/.*
- **Manual**: https://www.bioinformatics.babraham.ac.uk/projects/fastqc/
- **FASTQC explained**: https://www.bioinformatics.babraham.ac.uk/projects/fastqc/


FastQC is a program to summarize read qualities and base composition. Since we have millions of reads there is no practical way to do this by hand. We call the program to parse through the fastq files and determine some basic statistics about the data. The input to the program is one or more fastq file(s) and the output is an html file with several figures. The link above describes what each of the output figures are describing. I mainly focus on the first graph which visualizes our average read qualities and the last figure which shows the adapter content. Note that this program does not do anything to your data, it merely reads it and writes a report.

In [46]:
%%bash
# make a directory to store the output
mkdir -p output-fastqc

# run the program on the two read files (R1 and R2).
echo "RUNNING FASTQC"
fastqc raw-reads/*.fastq.gz -o output-fastqc --threads 24 > /dev/null 2>&1

# the resulting folder should contain a zipped archive and an html file, we can ignore the zipped archive which is redundant.
ls output-fastqc/

RUNNING FASTQC
SRR10056829_1_fastqc.html
SRR10056829_1_fastqc.zip
SRR10056829_2_fastqc.html
SRR10056829_2_fastqc.zip
null_fastqc.html
null_fastqc.zip
Process is interrupted.


In [44]:
## view HTML report
from IPython.display import IFrame
IFrame('output-fastqc/SRR10056829_1_fastqc.html', width=1000, height=550)

## Step 2: Quality and adapter trimming with FASTP

- **Program**: **fastp** - An Ultra-Fast All-in-One FASTQ Preprocessor
- **Citation**: *Chen, S. (2023). Ultrafast one‐pass FASTQ data preprocessing, quality control, and deduplication using fastp. Imeta, 2(2), e107.*
- **Manual**: [fastp GitHub Repository](https://github.com/OpenGene/fastp)
- **Conda**: [fastp on Bioconda](https://anaconda.org/bioconda/fastp)


Description:
**fastp** is an ultra-fast, all-in-one tool designed for preprocessing and quality control of high-throughput sequencing data. It performs several critical tasks, including base correction, adapter trimming, read filtering, and quality score recalibration. Additionally, it provides deduplication to remove PCR duplicates and offers comprehensive quality control through detailed reports and visualizations. Its efficiency allows it to handle large datasets rapidly, making it ideal for processing next-generation sequencing (NGS) data in various formats like FASTQ. Alternative tools for similar purposes include **Trimmomatic**, **Cutadapt**, and **FASTX-Toolkit**, which offer some of the same functionalities but may differ in speed, ease of use, and specific features.

In [45]:
%%bash

# make a directory to store the output
mkdir output-fastp

# define input variables
forward=raw-reads/SRR10056829_1.fastq.gz
reverse=raw-reads/SRR10056829_2.fastq.gz

# run the program
fastp --in1 $forward --in2 $reverse --out1 output-fastp/trimmed_1.fastq.gz --out2 output-fastp/trimmed_2.fastq.gz

# the resulting folder should contain new trimmed read files
ls output-fastp/

mkdir: cannot create directory ‘output-fastp’: File exists
Read1 before filtering:
total reads: 1591280
total bases: 159128000
Q20 bases: 157594869(99.0365%)
Q30 bases: 152891677(96.0809%)

Read2 before filtering:
total reads: 1591280
total bases: 159128000
Q20 bases: 155820667(97.9216%)
Q30 bases: 150621528(94.6543%)

Read1 after filtering:
total reads: 1568979
total bases: 154244678
Q20 bases: 153095395(99.2549%)
Q30 bases: 148904585(96.5379%)

Read2 after filtering:
total reads: 1568979
total bases: 154244678
Q20 bases: 152644379(98.9625%)
Q30 bases: 147803821(95.8243%)

Filtering result:
reads passed filter: 3137958
reads failed due to low quality: 43808
reads failed due to too many N: 794
reads failed due to too short: 0
reads with adapter trimmed: 214702
bases trimmed due to adapters: 5320114

Duplication rate: 2.70292%

Insert size peak (evaluated by paired-end reads): 168

JSON report: fastp.json
HTML report: fastp.html

fastp --in1 raw-reads/SRR10056829_1.fastq.gz --in2 raw-re

trimmed_1.fastq.gz
trimmed_2.fastq.gz


### Explanation of fastp Results

The **fastp** results printed above show the quality of the sequencing data both before and after filtering. These results provide insight into the number of reads, base quality, and the effectiveness of filtering steps like adapter trimming and quality control.

- **<u>Before filtering</u>**:
  - **Read1** and **Read2** both start with 1,591,280 reads, with high overall quality scores. Over 99% of bases have a quality score (Q20) above 20, and about 96% of bases have a quality score (Q30) above 30, indicating good overall quality in the raw data.

- **<u>After filtering</u>**:
  - The total number of reads decreases slightly to 1,568,979 for both **Read1** and **Read2**. Despite this reduction, the quality improves, with over 99% of bases achieving Q20 and over 96% of bases achieving Q30 for **Read1**, and over 98% for **Read2**. This shows that filtering removed low-quality reads while maintaining high-quality data.

- **<u>Filtering summary</u>**:
  - A total of 31,379,958 reads passed the quality filter, while 43,808 reads failed due to low quality. A small number (794) failed due to the presence of too many 'N' bases, and no reads were discarded due to being too short.
  - **Adapter trimming** removed 5,320,114 bases across both reads, which improves the accuracy of downstream analysis.

- **<u>Duplication rate</u>**: 
  - The duplication rate is low at 2.70%, meaning most reads are unique and useful for assembly and analysis.

- **<u>Insert size</u>**: 
  - The insert size peak is 168, which reflects the typical distance between paired-end reads, providing valuable information about the sequencing library.

These results suggest that **fastp** was effective in cleaning the raw sequencing data by removing low-quality reads and trimming adapter sequences, resulting in improved data quality suitable for downstream analyses.

## Step 3: Genome Assembly

- Program: **SPAdes - St. Petersburg genome assembler**
- Citation: *Bankevich, A., Nurk, S., Antipov, D., Gurevich, A. A., Dvorkin, M., Kulikov, A. S., ... & Pevzner, P. A. (2012). SPAdes: a new genome assembly algorithm and its applications to single-cell sequencing. Journal of computational biology, 19(5), 455-477.*  
- Manual: https://ablab.github.io/spades/  

SPAdes (St. Petersburg genome assembler) is a genome assembly algorithm which was designed for single cell and multi-cells bacterial data sets. However, it might not be suitable for large genomes projects. SPAdes works with Ion Torrent, PacBio, Oxford Nanopore, and Illumina paired-end, mate-pairs and single-end reads.

<div class="alert alert-block alert-info"><b>Tip</b>: I recommend running the code below, then watching the following video tutorial or reading more while it runs.</div>

In [51]:
%%bash
## Assemble trimmed reads using the SPAdes assembler.

# define input
forward=output-fastp/trimmed_1.fastq.gz
reverse=output-fastp/trimmed_2.fastq.gz

# assemble reads with SPAdes
echo "RUNNING THE SPADES PIPELINE, PLEASE REVIEW MATERIALS BELOW WHILE IT RUNS"
spades.py -1 $forward -2 $reverse -o output-spades --threads 24 > logfile_spades.txt 2>&1


RUNNING THE SPADES PIPELINE, PLEASE REVIEW MATERIALS BELOW WHILE IT RUNS


## Video Lesson: How genome assembly of short reads works.

The following video describes the process in detail. Image credit. https://homolog.us/Tutorials/book4/p2.1.html

[![Sequencing by Synthesis](images/debruijn.png)](https://www.youtube.com/watch?v=ZmF6QROPlTU)

## Video Recap

### De Bruijn Graphs in Genome Assembly

A **de Bruijn graph** is a structure used in genome assembly to represent sequences using **k-mers**. Here’s how it works:

1. **K-mers**: These are short sequences of length *k* derived from a longer DNA sequence. For example, from the sequence "AGCT", the 2-mers are "AG", "GC", and "CT".

2. **Graph Construction**: Each k-mer becomes a vertex in the graph. An edge is drawn from k-mer **A** to k-mer **B** if the last *k-1* nucleotides of **A** match the first *k-1* nucleotides of **B**. This creates a directed graph where paths through the vertices represent potential sequences in the genome.

3. **Contigs and Scaffolds**:
   - **Contigs**: These are contiguous sequences formed by traversing paths in the graph.
   - **Scaffolds**: Longer constructs that may combine multiple contigs, providing additional structure to the genome assembly.

De Bruijn graphs enable efficient handling of sequencing data, especially for complex genomes with repetitive sequences.
  
  
### Key challenges for genome assembly:
**<u>Intrinsic Challenges:</u>**
1. **Heterozygosity:** The alleles of a gene are not he same, yet we force them into a single consensus sequence.
2. **Paralogy vs. Alleleism**: Genes come from other genes by a process of duplication. This results in two or more similar genes in an organism. There are two alleles in a diploid organism that are very similiar. How do you tell a duplicated gene from alleles of a gene?
3. **Sequence complexity**: Simple sequence repeats (SSR), large scale repeats like transposable elements (TEs).

**<u>Extrinsic Challenges:</u>**
1. **Quality of DNA sequences (sequencing errors):** Each sequencing technology has specific patterns of error.
2. **Length of DNA sequence reads**: Shorter reads are less likely to be unique or to include unique K-mers (see below).
3. **Coverage**: Depth of coverage is a random process at best. Consequently some regions of the genome will have low levels of coverage.
4. **Memory intensive**: Inherently requires large amounts of RAM for assebly and storage of input and output.




## Conclusion and wrap up

End of submodule 1.  After learning about how genome sequencing works (and where it can go wrong) we went through standard processing steps (FASTQC and FASTP) and ultimetly ran a genome assembly tool.

The final output from spades represents our draft genome sequence.

Be sure to shut down your instance. We will run one more bit of code to copy our genome file for use in submodule 2

In [ ]:
%%bash
# did the assembler finish?
ls output-spades/

In [2]:
%%bash
# make a copy of the genome sequence in a new directory. This data will be used for Submodule 2.

mkdir -p assembled-genome
cp output-spades/contigs.fasta assembled-genome/